content/blog/2019-09-25-interpreter-vector-support/index.md

```diff
@@ -0,0 +1,134 @@
++++
+title = "Vector Instruction Support in the Bril Interpreter"
+extra.bio = """
+  Philip Bedoukian is a 3rd year PhD student in ECE. His research focuses on reconfigurable hardware.
+"""
+extra.author = "Philip Bedoukian"
++++
+
+## Problem Statement
+
+The Bril interpreter does not take advantage of the vector instructions available in modern CPUs. This presents two problems: the first is that a Bril backend cannot generate vector instructions because they are not present in the intermediate language. The second issue is that  interpreting vector instructions is slow because it requires a loop over each vector element. We propose to support vector instructions in the interpreter and accelerate the interpretation of them using intrinsic vector instructions.
+
+## Vector Instructions
+
+Single-Instruction Multiple-Data (SIMD) allows computer hardware to obtain signifigant speedups over the conventional execution paradigm, Multiple-Instruction Multiple-Data (MIMD). SIMD executes fewer instructions than MIMD to accomplish the same amount of work. In SIMD, multiple arithmetic operations are grouped together under one instruction. Generally, this aritmetic is executed over multiple execution units at the same instant (spatial SIMD), although some implementations allow the same instruction to use the same functional unit over multiple cycles (temporal SIMD). Various hardware architectures allow the programmer to support SIMD by exposing vector instructions in their ISA. Typically, there is a vector register file which holds multiple elements per register and can participate in vector arithmetic.
+
+## Vector Support in the Bril Interpreter
+
+### Memory Support
+
+Vector instructions are only useful when operating on large chunks of data. Conversely, registers are not designed to hold this amount of data. A prerequisite for vector instructions is support for data memory in the interpreter. The interpreter memory is implemented like a stack. The program is allowed to access any location in the fixed size memory. Stack frames were not implemented because the interpreter only supports a single function.
```

Comment:
I'm not quite sure what you mean by this:

> The interpreter memory is implemented like a stack.

It sounds like this is actually like a big C heap, i.e., a flat address space you can load and store from? It doesn't seem much like a stack, even ignoring the fact that there are no function calls.